Shallow renderer and test utils bundles

bvaughn opened the pull request:
Adds new bundles introduced with React 15.5 release to master (and 16 alpha):

### react-dom/test-utils

This new bundle contains what used to be `react-addons-test-utils`. This bundle shares things from `react-dom` rather than duplicates them.

A temporary `createRenderer` method has been left behind as a way to access the new shallow renderer. This is for the ReactNative release cycle only and should be going away before the final release.

### react-test-renderer/shallow

This new shallow renderer is almost entirely stand-alone (in that it doesn't use the React reconciler or scheduler). The only touch points are `ReactElement` and prop/context validation. This renderer is stack and fiber compatible.

PS I used `__SECRET_INTERNALS` please don't fire me 😛

gaearon commented on src/isomorphic/classic/element/ReactElement.js:
```diff
@@ -162,6 +162,13 @@ var ReactElement = function(type, key, ref, self, source, owner, props) {
+        enumerable: false,
+        writable: false,
+        value: owner,
+      });
```
Related past discussion https://github.com/facebook/react/pull/6355

gaearon replied:
Can you verify with @sebmarkbage he has no concerns here?

bvaughn replied:
Seb and I chatted about this idea in person. He didn't seem to, but I'll let him comment here (or I can follow up with him in person on Monday). Sure!

bvaughn commented:
Adding a few more folks who may be interested in looking over this change

gaearon commented on packages/react-test-renderer/shallow.js:
```diff
@@ -0,0 +1,7 @@
+
+if (process.env.NODE_ENV === 'production') {
+  throw Error('shallow renderer is not available in production mode.');
+} else {
```
Should we also throw for the main test renderer entry point then?

oguzgelal replied:
@gaearon @bvaughn I'm having a problem with these lines: I'm using [enzyme](https://github.com/airbnb/enzyme) and it uses shallow rendering, _for tests_, while `NODE_ENV` is set to `production`. So my tests are failing because of this.

gaearon commented on packages/react-test-renderer/shallow.js:
```diff
@@ -0,0 +1,7 @@
+
+if (process.env.NODE_ENV === 'production') {
+  throw Error('shallow renderer is not available in production mode.');
+} else {
```
Or should we create prod versions of all of them? I don't see benefit but somebody might come up with a use case. Maybe just worth keeping in mind.

bvaughn replied:
> Or should we create prod versions of all of them?

I don't see the value of this, and it would cause problems with `__DEV__` only things like the non-enumerable `_owner` attribute.

gaearon commented on scripts/rollup/results.json:
```diff
@@ -122,20 +122,40 @@
+      "gzip": 55887
+    },
+    "react-test-utils.development.js (NODE_DEV)": {
+      "size": 510240,
```
I think this file was deleted so the stat is probably not needed?

gaearon commented on src/renderers/testing/ReactShallowRenderer.js:
```diff
@@ -0,0 +1,121 @@
+      const tree = this._renderer.toTree();
+      if (tree && tree.rendered) {
+        // If we created a context-wrapper then skip over it.
+        const element = tree.type.childContextTypes
```
This seems fragile. What if it isn't *our* context but the root element just happens to be a context provider?

To be honest I'm not a fan of second undocumented `context` argument. Our other APIs don't take it. I think it's remaining from the time context was owner-based and we might as well deprecate it and tell people to use their own context providers explicitly. But it's probably annoying to do in this PR.

Let's just be stricter with checking here.

gaearon commented on src/renderers/testing/ReactShallowRenderer.js:
```diff
@@ -0,0 +1,121 @@
+
+        // Convert the rendered output to a ReactElement.
+        // This supports .toEqual() comparison for test elements.
+        return React.createElement(
```
Would this re-trigger any validation? Would people see duplicate messages?

bvaughn replied:
This would only re-trigger if the `ReactElement` returned from the top-level render was a class component with `propTypes`. In that case, it still wouldn't show a duplicate message because `checkPropTypes` tracks `loggedTypeFailures` and avoids re-logging duplicates.

Do you think this is reasonable? Should I look into ways of bypassing `checkPropTypes` for this case? (Or something else?)

gaearon replied:
What about the key warning?
It’s probably fine, just making sure you’re aware of this.

bvaughn replied:
Key warning will only log once as well.

Yup, much appreciated for the double check. 😄

gaearon commented on src/renderers/testing/ReactShallowRenderer.js:
```diff
@@ -0,0 +1,121 @@
+
+  getRenderOutput() {
+    if (this._renderer) {
+      const tree = this._renderer.toTree();
```
This will likely make all shallow tests slower. Would be good to measure it on some mid size codebase to check the impact. It might be better to add some way of bailing out if it's too bad.

gaearon commented on src/test/ReactTestUtils.js:
```diff
@@ -12,23 +12,27 @@
 'use strict';
 
```
While we're at it can we move `src/test` to a more appropriate place (considering the package structure)?

bvaughn replied:
Done!

gaearon commented on src/test/__tests__/ReactTestUtils-test.js:
```diff
@@ -19,286 +22,290 @@ var ReactTestUtils;
-    class SomeComponent extends React.Component {
-      render() {
+  // Shallow renderer only implemented for Fiber in 16+
+  if (ReactDOMFeatureFlags.useFiber) {
```
I'm a bit confused by this comment. Can we branch on require instead? Otherwise it looks like we stopped testing shallow renderer in Stack at all. However the Stack version is what we're currently running in www.

bvaughn replied:
> Otherwise it looks like we stopped testing shallow renderer in Stack at all.

This new shallow renderer _is_ fiber only. Seemed like too much effort to create a new one that supports stack and works with our new bundles- just to throw it away. (I mentioned this in the description but maybe it wasn't prominent enough.)

I'm not sure how to address www yet.

gaearon replied:
Oh. I didn't realize we *deleted* the Stack based one. Our usual strategy is to keep both until we’ve migrated everything that uses the old one (e.g. www). This lets us revert to using the old one if the new one has issues. Otherwise once this is merged, we have no way but to fix forward. Which might be fine, but it’s blocking any future syncs, and I don’t know if this will pass www tests or not.

Usually we keep both, and let the tests pick the right one based on the flag.

bvaughn replied:
Yeah, totally understood- and I'm glad you bought it up. I'll figure out a strategy for www tests with stack once the fiber bundles and renderer is solid. 😁

gaearon commented:
I am also concerned that running `toTree()` will also make shallow rendering more fragile. For example imagine a leaf component changes to throw in `render()`. Now your grandparent shallow rendering test fails even though it's meant to be isolated. Am I missing something here?

bvaughn commented:
Great point Dan.

I just realized something silly anyway 😐 The new renderer actually causes _render_ to be called on the full tree because of how the reconciler works. I need to think about this a bit more.

gaearon commented:
Yea, this is what I meant by the "bailout" in the comment above. The original PR by @lelandrichardson contained this logic in a more generic way (it allowed "redirecting" types) and I think @sebmarbage had a strategy for using this for shallow renderer in https://github.com/facebook/react/pull/8931#issuecomment-278821673. This way it would only render one level deep.

bvaughn commented:
Gotcha! Thanks for pointers. 😄

bvaughn commented:
Gave this some more thought over the weekend and tried a couple of alternate implementations. One was similar to `ReactNoopRenderer` but required monkey-patching `ReactCurrentOwner` to work which felt too hacky. The other one (which I have now committed+pushed) is stand-alone. I want to do a little more testing to make sure I didn't introduce a regression WRT the lifecycle hooks or something similar- but I feel this route is probably better than the previous test-renderer-based approach.

bvaughn commented:
I ran shallow renderer tests in www against the new `ShallowRenderer`. There aren't that many of them unfortunately (only 64) but all pass except:
* A few that also use Enzyme
* `P2PMemoField.react-test.js` which uses internals (`renderer._instance._instance`). I'll fix this later as part of syncing the new shallow renderer to www. 
* `P2PProductItemChatThreadBanner.react-test.js` which uses `ReactLayeredComponentMixin_DEPRECATED`. Specifically this one fails because it calls `ReactDOM.renderSubtreeIntoContainer` which errors because `ReactInstanceMap` doesn't contain the shallow instance. I think it's reasonable to not support this test with the new renderer- particularly since we were hoping to decouple the shallow renderer from DOM.

I also tried plugging the new renderer into a local copy of Enzyme but didn't spend much time on it once I realized there was some internal coupling (eg `ShallowWrapper.instance = () => this.renderer._instance ? this.renderer._instance._instance : null`). I believe the Enzyme team has plans to implement their own, Enzyme-compatible shallow renderer for the next major version so I think it's okay to leave this off for now.

koba04 commented on src/renderers/testing/ReactShallowRenderer.js:
```diff
@@ -0,0 +1,216 @@
+
+  if (typeof instance.shouldComponentUpdate === 'function') {
+    if (instance.shouldComponentUpdate(props, state, context) === false) {
+      return rendered;
```
Should React update `instance.props` even if SCU returns `false`?

bvaughn replied:
Good catch, @koba04.

koba04 replied:
In addition to that, should shallow renderer update `state` and `context` as well?

gaearon commented:
Is there a particular reason you didn't reuse `ReactFiberClassComponent` helpers but inlined a version of them?

gaearon commented on src/renderers/testing/ReactShallowRenderer.js:
```diff
@@ -100,26 +59,146 @@ class ReactShallowRenderer {
+  const rendered = instance.render();
+
+  if (typeof instance.componentDidMount === 'function') {
+    instance.componentDidMount();
```
Do we actually want to run this in shallow mode then? I would expect that this would crash if you use DOM refs. I don't think we used to fire it in 15.

bvaughn replied:
Fair enough. Although we have tests for some lifecycle hooks (eg `componentWillUnmount`) so it seems weird not to have cDU. I guess you're right though, that without access to host refs it's a bit weird to call this lifecycle hook.

gaearon replied:
Yes, "in theory" it would be correct to call it but I don't think it would work in practice. Curious how this doesn't blow up www.

gaearon replied:
cc https://github.com/facebook/react/issues/7678 and "`componentDidServerRender`". This is a similar problem. We can't currently separate DOM flush lifecycle from a more generic "completion" lifecycle.

bvaughn replied:
We actually don't have that many uses of shallow renderer in www tests.

I'll remove cDU though and add an inline comment about why we do it!

gaearon replied:
I think if we aren't ready to do an intentional decision here, we should just maintain existing behavior (I'm not sure what the existing behavior is). We should also add tests for this behavior.

bvaughn replied:
Yup! I had already added lifecycle hook tests in this PR. 😄 

I'll update to match the fact that we _won't_ call cDM.

gaearon replied:
It would be good to run them on existing Stack implementation to verify tests match previous behavior.

gaearon commented on src/renderers/testing/ReactShallowRenderer.js:
```diff
@@ -100,26 +59,146 @@ class ReactShallowRenderer {
+  const rendered = instance.render();
+
+  if (typeof instance.componentDidUpdate === 'function') {
+    instance.componentDidUpdate(oldProps, oldState, oldContext);
```
Ditto.

gaearon replied:
Let's not pass `prevContext`?

bvaughn replied:
Pretty sure I saw that we had tests internally that depend on cDU being called for shallow rendered components.

What's the concern with calling this?

gaearon replied:
Same as `componentDidMount`: it assumes DOM is there and could operate on refs. I guess at this point we need a redesign 😢 . Let's just do whatever 15 version does, and we can look at it again in 17 now that it's decoupled. You could also check with @sebmarkbage if he had short-term thoughts here.

bvaughn replied:
Good call.

I recall now that it was while testing Enzyme that I noticed the expectations that `componentDidUpdate` will be called.

Enzyme has a [spec](https://github.com/airbnb/enzyme/blob/a9a3627c3b4bf877fd0062379b1dac6344c2b69b/test/ShallowWrapper-spec.jsx#L2722) that verifies previous shallow renderer behavior. It shows that `shouldComponentUpdate

koba04 replied:
According to my last research, ShallowRenderer calls cDU with `setState`, but it isn't called with `render`.

You can see the result at `lifecycleExperimental` being false in this link.
(Enzyme's setProps and setContext are just calling render again)

* https://github.com/airbnb/enzyme/pull/318#issuecomment-225148673

bvaughn replied:
Hm, that's interesting. It's a little confusing to trace through their tests because of how they're structured. `it` blocks often depend on state initialized by a parent `describe` block so it takes a bit of scanning back and forth, and there's a couple of flags (eg `lifecycleExperimental`, `disableLifecycleMethods`) that change things.

I believe you're right though. I'll try to mimic the old shallow renderer behavior in this regard, even though I don't understand the reasons for it. 🙁

koba04 replied:
@bvaughn 
I guess the reason is that `ShallowRender.render` run on an outside transaction, which means that enqueued functions have never been called.

* https://github.com/facebook/react/blob/master/src/test/ReactShallowRenderer.js#L68
* https://github.com/facebook/react/blob/master/src/renderers/shared/stack/reconciler/ReactCompositeComponent.js#L1043-L1053

I used to create a PR to fix it. I hope it can be of any help to you.

* https://github.com/facebook/react/pull/7912

gaearon commented on src/renderers/testing/ReactShallowRenderer.js:
```diff
@@ -100,26 +59,146 @@ class ReactShallowRenderer {
+    } else {
+      const prototype = element.type.prototype;
+
+      if (typeof prototype.render === 'function') {
```
Can we keep the way we determine class vs function consistent? This seems like an ad-hoc way that doesn't match how we do this elsewhere (`prototype.isReactComponent`). I'm also not sure if this supports "factory" components (`function Foo() { return { render() { return <div /> } } }`).

bvaughn replied:
Sure. We aren't super consistent with how we check. This ^ was just mimicking the way we check in a few places I saw- but if `prototype.isReactComponent` is the more recommended way I'll switch. :)

gaearon replied:
I am only aware of these two places:

* Fiber https://github.com/facebook/react/blob/master/src/renderers/shared/fiber/ReactFiber.js#L234
* Stack  https://github.com/facebook/react/blob/master/src/renderers/shared/stack/reconciler/ReactCompositeComponent.js#L45

Are there more?

bvaughn replied:
Those places do it the way you mentioned- which is great, I'm happy to change to that. 😁 

There's a few places we're check for a `render` function too but they're probably just older:
* [`ReactFiberBeginWork`](https://github.com/facebook/react/blob/10eaf26809bdb1874e14f0879190c356937c2e3e/src/renderers/shared/fiber/ReactFiberBeginWork.js#L517-L519)
* [`reactComponentExpect`](https://github.com/facebook/react/blob/10eaf26809bdb1874e14f0879190c356937c2e3e/src/test/reactComponentExpect.js#L189-L190)
* [`findNodeHandle`](https://github.com/facebook/react/blob/10eaf26809bdb1874e14f0879190c356937c2e3e/src/renderers/native/findNodeHandle.js#L103)
* [`findDOMNode`](https://github.com/facebook/react/blob/10eaf26809bdb1874e14f0879190c356937c2e3e/src/renderers/dom/shared/findDOMNode.js#L70)

gaearon replied:
The first one tests for *factory components*. This is a bit different. See my example in the comment above. We support components that are functions but return instances.

Other examples depend on instances already existing. I was talking about how to determine something should be `new`'d *before it gets created*. It is important because it is unsafe to `new` some things (for example arrow functions), and because we support a pattern that doesn't need `new` but still returns instances (factory components).

bvaughn replied:
Good clarification.

gaearon replied:
Sorry, I'm on the road and probably not making sense. What I'm trying to say is that the other examples check for something on the instance. But instance doesn't exist when you're "choosing" how to create it. There are only two places where we "choose" that (based on `type` alone), and these are the places we need to be consistent with since we also have just the `type`.

The overall algorithm is:

1. Check for prototype.isReactComponent.
2. If it exists, call type with new.
3. If it doesn't, call type without new.
4. Look at what type returned. If it's an object with render method, treat it as class instance from now on. Otherwise treat it as a functional component that returned an element.

bvaughn replied:
No no. Your previous comment totally made sense. 😁

gaearon commented on src/renderers/testing/ReactShallowRenderer.js:
```diff
@@ -12,78 +12,37 @@
-  React.Children.toArray(children);
+const {
+  ReactDebugCurrentFrame,
+} = React.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED;
```
Shouldn't `const { ReactDebugCurrentFrame } = require('ReactGlobalSharedState')` work? If not we need to fix that.

bvaughn replied:
Ah, sure, looks like it should. I wasn't aware of that.

bvaughn commented:
> Is there a particular reason you didn't reuse ReactFiberClassComponent helpers but inlined a version of them?

The helpers depend on having Fibers (which shallow renderer doesn't use) and having access to other internals that I was hoping to avoid having shallow renderer depend on.

gaearon commented:
I don't know what's the deal with GH collapsing my comments but please check a few more of my comments that shouldn't be "outdated" 😄

bvaughn commented:
> I trust you to fix issues in comments above and to address TODOs you currently have in the code. But let's make sure this works in other third party projects (e.g. would be good to run this against at least a few third party projects that use shallow rendering).

Thanks Dan! That's why I polled on Twitter for a few external projects. Enzyme has been informative, although I won't be able to really verify that the new shallow renderer works with it since there are so many hooks to internals.

I'll circle back WRT the TODO comments after the upcoming alpha makes its way into RN and we can actually get rid of the deprecated properties. The coordination between React + ReactNative makes this a bit tricky.

bvaughn commented:
> (e.g. would be good to run this against at least a few third party projects that use shallow rendering)

All 617 Enzyme unit tests pass for react/react-dom 15 + the new shallow renderer if I make the following patch:
```diff
diff --git a/src/ShallowWrapper.js b/src/ShallowWrapper.js
index bd23106..ade4e73 100644
--- a/src/ShallowWrapper.js
+++ b/src/ShallowWrapper.js
@@ -194,7 +194,11 @@ class ShallowWrapper {
     if (this.root !== this) {
       throw new Error('ShallowWrapper::instance() can only be called on the root');
     }
-    return this.renderer._instance ? this.renderer._instance._instance : null;
+    const instance = this.renderer._instance;
+
+    return instance && instance._instance
+      ? instance._instance
+      : instance;
   }
 
   /**
```

The above patch accounts for the fact that we no longer auto-wrap functional components in a class component.

Enzyme isn't compatible with React 16 yet because there are a lot of hooks into internals. But this improves my confidence about the new shallow renderer approach.

gaearon commented:
To be clear I don't think it's as important that *Enzyme's* own tests pass (which may rely on internals too much) since it's bound for major anyway. But we should test it on large projects *using* shallow rendering (via Enzyme or not).

bvaughn commented:
Uh, right. But Enzyme's own tests seem like a pretty useful metric given their volume. If the new renderer passes all of their tests- that's a good sign, no?

I'm having trouble testing it on projects _using_ shallow renderer because none of them have updated to be React 16 compatible yet.

koba04 commented:
What do you think ShallowRender has a new API like `getPublicInstance`, which returns a public instance?
Currently, Enzyme depends on React internal to get a public instance.

* https://github.com/airbnb/enzyme/blob/master/src/ShallowWrapper.js#L197

It would be nice if it is accomplished without depending on React internals.
To do that, Enzyme needs some works because Enzyme is using a public instance even if it's a Stateless Functional Components, but it can be fixed.

bvaughn commented:
> What do you think ShallowRender has a new API like getPublicInstance, which returns a public instance? Currently, Enzyme depends on React internal...

Yeah, I had to patch `ShallowWrapper` to test it with the new shallow renderer. :grimace:

AirBnb has plans to rewrite Enzyme internals for the next release though. The idea is for libraries like React (and similar ones- Preact, Inferno, etc) to provide an abstraction layer / adapter that Enzyme consumes rather than reaching into internals. The Enzyme team can then create adapters on their side (eg one for each major version of React) so their core can be cleaned up. ([This file](https://github.com/airbnb/enzyme/blob/d49254b42d5764043ecf9e9ce914815f11c199d3/docs/future/compatibility.md) has since been deleted, or maybe moved, but I think it explains the general idea.)

That's a bit orthogonal to this PR though probably. At the moment, I don't think they have any plans of actually using this new shallow renderer. It's for non-Enzyme users (including Facebook's non-Enzyme tests).

bvaughn commented:
By the way, Dan- thanks for the unblock approval. I'm going to spend a little more time walking through the 30 failing tests no www and writing down the reasons why they're failing (patching things up if possible) to make our later sync easier. I hope to merge this PR sometime tomorrow.

gaearon commented:
Sounds great!

bvaughn commented:
All www tests pass now! Added a few more tests for edge-cases I found while running through www tests. Soon as Circle gives it a green check, I'll merge.

bvaughn commented:
For posterity (or future-me) here's the list of mods I made in www while testing:

1. Add 'prop-types' and 'create-react-class' packages to `html/shared/node_modules`
2. Add magicdirs entries for the new bundles and the above NPM packages so Haste could find them.
3. Patch Enzyme's `ShallowWrapper.instance()` slightly to account for the fact that the new renderer doesn't wrap function components in a class component.
4. Noop mock `unstable_renderSubtreeIntoContainer` since any calls to this method would fail due to dependencies on `ReactInstanceMap` which isn't used by the new shallow renderer. (This was required for tests like `P2PProductItemChatThreadBanner.react-test.js` that use `ReactLayeredComponentMixin_DEPRECATED`.)
5. Updated `ReactDOMServerStack-dev` to read `checkPropTypes` from `prop-types/checkPropTypes` rather than `React.checkPropTypes`. (Enzyme invokes this using `renderToStaticString`.) Also updated `ReactARTStack-dev` in a similar fashion.
6. Replaced `renderer._instance._instance` with `renderer.getMountedInstance()` in `P2PMemoField.react-test`
7. Mock `XControllerURIBuilder.prototype.setInt.mockReturnThis()` for `TestConsoleLoader-test.react.js`

I'll submit a WIP diff for the above changes internally as well for when we eventually sync a newer alpha to www.

koba04 commented:
Great work! 🎉

bvaughn commented:
Thanks so much for the helpful info you provided @koba04 !

gaearon commented:
Great work!

gaearon commented:
I added prod builds in 16.1.  You can try `16.1.0-beta` now.

oguzgelal commented:
Hmm. Upgraded to `16.1.0-beta` but still having the issue.

gaearon commented:
This code doesn't exist in it. 

https://unpkg.com/react-test-renderer@16.1.0-beta/shallow.js

If you still see the issue you have not upgraded. Or forgot to update some packages.

oguzgelal commented:
Ah you're right, yes I updated `react` not `react-test-renderer`. Thanks!
